src/ray/rpc/node_manager_server.h

```diff
@@ -25,33 +25,31 @@ class NodeManagerServiceHandler {
                                  RequestDoneCallback done_callback) = 0;
 };
 
-/// The `GrpcServer` for `NodeManagerService`.
-class NodeManagerServer : public GrpcServer {
+/// The `GrpcService` for `NodeManagerService`.
+class NodeManagerGrpcService : public GrpcService {
  public:
   /// Constructor.
   ///
-  /// \param[in] port See super class.
-  /// \param[in] main_service See super class.
+  /// \param[in] io_service See super class.
   /// \param[in] handler The service handler that actually handle the requests.
-  NodeManagerServer(const uint32_t port, boost::asio::io_service &main_service,
-                    NodeManagerServiceHandler &service_handler)
-      : GrpcServer("NodeManager", port, main_service),
-        service_handler_(service_handler){};
+  NodeManagerGrpcService(boost::asio::io_service &io_service,
+                         NodeManagerServiceHandler &service_handler)
+      : GrpcService(io_service), service_handler_(service_handler){};
 
-  void RegisterServices(grpc::ServerBuilder &builder) override {
-    /// Register `NodeManagerService`.
-    builder.RegisterService(&service_);
-  }
+ protected:
+  grpc::Service &GetGrpcService() override { return service_; }
 
   void InitServerCallFactories(
+      const std::unique_ptr<grpc::ServerCompletionQueue> &cq,
```

Comment:
just pass in a reference of the CQ? because cq can't be nullptr here.

Reply:
With that calling `ServerBuilder's` `RegisterService` needs to be moved to after `AddCompletionQueue`, from the current grpc implementation it should be ok.

Reply:
But the `protobuf` generated C++ code requires the service rpc method to have a pointer of `CompletionQueue` as a param, thus changing it to a reference would not work here:)

Reply:
gRPC's api requires a raw pointer as the param. So passing reference can still work?

Reply:
Looking at the code again, I think I overlooked it. Reference doesn't work here.
It has to be `a reference to a unique_ptr` instead of a reference, because the `GrpcServer's` `RegisterService` is called **before** the `Run` function is called, thus when registering a service to grpc server, `cq_` is not initialized yet.

Reply:
I see. thanks!